Add Show, Reviewed, and Priority commands

g-lilian opened the pull request.

g-lilian committed:
Implement ShowCommand

g-lilian committed:
Implement ReviewedCommand

Users can mark a flashcard as reviewed by inputting reviewed [index]. The review status is displayed beside the flashcard name when 'list' command is called.

g-lilian committed:
Implement PriorityCommand

Users can now set the priority of a task using priority [taskIdx] [priorityLevel]. Exception handling has not been fully implemented for this.

quinnyyy commented on src/main/java/seedu/tp/commands/ShowCommand.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public void execute() throws InvalidFlashcardIndexException {
+        ui.showFlashcard(flashcardList, index);
```
I would personally refactor this to be something like:
Flashcard flashcard = flashcardList.getFlashcardAtIndex(index);
ui.showFlashcard(flashcard);

That way the ui.showFlashcard() method has one responsibility

g-lilian replied:
Edited, thanks.

quinnyyy commented on src/main/java/seedu/tp/parser/Parser.java:
```diff
@@ -63,9 +65,15 @@ public Command parseCommand(String userInput) throws HistoryFlashcardException {
+        case SHOW_COMMAND:
+            try {
+                return new ShowCommand(flashcardList, Integer.parseInt(splitInput[1]), ui);
+            } catch (Exception e) {
```
I think it's generally a better idea to be explicit about what Exception we are catching, right?

g-lilian replied:
Noted and changed to IndexOutOfBoundsException.

quinnyyy commented on src/main/java/seedu/tp/commands/ReviewedCommand.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    public void execute() throws InvalidFlashcardIndexException {
+        try {
+            flashcardList.getFlashcardAtIdx(index).setReviewStatus(true);
```
Think this is more readable if you do something like:
Flashcard flashcard = flashcardList.getFlashcardAtIdx(index);
flashcard.setReviewStatus(true);
ui.confirmFlashcardReview(flashcard);

g-lilian replied:
Edited, thanks.

quinnyyy commented on src/main/java/seedu/tp/flashcard/Flashcard.java:
```diff
@@ -52,4 +54,35 @@ public void setReviewStatus (boolean isReviewed) {
+     *
+     * @return "*"s to indicate priority level
+     */
+    public String getPriority() {
```
Maybe would be better idea to rename this to getPriorityAsString or something like that?

g-lilian replied:
Edited!

quinnyyy reviewed (changes requested):
looks good! left some comments about some small changes to consider

Zhilin-Huang commented on src/main/java/seedu/tp/flashcard/Flashcard.java:
```diff
@@ -9,11 +9,15 @@
     protected String summary;
     protected List<String> details;
+    protected boolean isReviewed;
+    protected int priorityLevel;
```
Consider using enum to define priority level? This way we can force priorityLevel to only take on a restricted set of values and we can also assign meaningful names to the priority levels

g-lilian replied:
Sure, do you have any names in mind? I can't think of how to make them more meaningful; first,second,third,fourth,fifth is more troublesome to type than 1,2,3,4,5.

Zhilin-Huang replied:
I'm thinking maybe we can use HIGH, MEDIUM, LOW, and also include DEFAULT

g-lilian replied:
So just 3 + 1 default levels instead of 5? The default currently is set to 'Not indicated' instead of the '***' display.

Zhilin-Huang replied:
I think it does not really matter for us we have 3 or 5 levels? 
Also we can have 3 levels for now and later if we think we need more, we can just add another enum field. 
For the default, we can have the `getPriorityAsString()` method to return the 'Not indicated' string

Zhilin-Huang reviewed (commented):
Generally looks good :)
But I assume that the text UI test needs to be updated because the code in the Ui class is changed
Also consider adding JUnit tests for the new command classes

g-lilian committed:
Refactor ShowCommand and ReviewedCommand

Small adjustments to make code more readable and organized.

quinnyyy commented on src/main/java/seedu/tp/commands/ReviewedCommand.java:
```diff
@@ -18,8 +19,9 @@ public ReviewedCommand (FlashcardList flashcardList, int index, Ui ui) {
         try {
-            flashcardList.getFlashcardAtIdx(index).setReviewStatus(true);
-            ui.confirmFlashcardReview(flashcardList.getFlashcardAtIdx(index));
+            Flashcard flashcard = flashcardList.getFlashcardAtIdx(index);
```
In ShowCommand it's index - 1. shouldn't they both be the same?

quinnyyy replied:
Other than this I think everything looks good.
Try to get the build passing, then I think it's good for merging

Zhilin-Huang replied:
> In ShowCommand it's index - 1. shouldn't they both be the same?

I'm thinking it might be better if we deal with the `-1` when we initialize all the commands. Like for `DeleteCommand` we have `return new DeleteCommand(flashcardList, Integer.parseInt(splitInput[1]) - 1);` in `Parser.java`, so maybe we can change `ShowCommand` to ` return new ShowCommand(flashcardList, Integer.parseInt(splitInput[1] - 1), ui);` and same for `ReviewedCommand`.

g-lilian replied:
Noted. I did that so that I didn't have to +1 when printing the index, but for consistency I agree it's better to deal with -1 in the Parser itself.

g-lilian committed:
Merge branch 'master' of https://github.com/AY1920S2-CS2113-T14-1/tp

g-lilian committed:
Refactor priorityLevel, ShowCommand and add comments

Made priorityLevel an enum with HIGH, MEDIUM, LOW and DEFAULT (not indicated) options. Added Javadoc comments to some command classes.

g-lilian committed:
Merge branch 'master' of https://github.com/AY1920S2-CS2113-T14-1/tp

g-lilian committed:
Refactor to pass checkstyle

g-lilian committed:
Refactor to pass checkstyle

g-lilian committed:
Edit text-ui-test

Edited to include Show, Priority, and Reviewed commands.
